File: engine/indicators.py
"""Candle metrics: RSI, ATR, BB, MACD, StochRSI, VWAP, S/R (v4.3)."""
from __future__ import annotations

import statistics
from typing import Any, Dict, List, Optional

from .util import to_float
from .scalper import wilder_rsi


def _ema(data: List[float], span: int) -> List[float]:
    if not data:
        return []
    alpha = 2 / (span + 1)
    out = [data[0]]
    for val in data[1:]:
        out.append(val * alpha + out[-1] * (1 - alpha))
    return out


def candle_metrics(klines: Any, label: str = "", warnings: Optional[List[str]] = None) -> Optional[Dict[str, Any]]:
    if not klines:
        return None
    if isinstance(klines, dict):
        raw = klines.get("data") or (klines.get("result") or {}).get("list") or []
    elif isinstance(klines, list):
        raw = klines
    else:
        return None
    if len(raw) < 5:
        return None

    rows = sorted(raw, key=lambda r: to_float(r[0]))
    open_times = [to_float(r[0]) for r in rows]
    if warnings is not None and len(open_times) != len(set(open_times)):
        warnings.append(f"candle_metrics({label}): duplicate open-time")
    if warnings is not None and len(rows) > 1:
        expected = to_float(rows[1][0]) - to_float(rows[0][0])
        for i in range(1, min(20, len(rows))):
            actual = to_float(rows[i][0]) - to_float(rows[i - 1][0])
            if expected > 0 and actual > 0 and abs(actual - expected) / expected > 0.1:
                warnings.append(f"candle_metrics({label}): gap ({actual:.0f} vs {expected:.0f})")
                break

    closes = [to_float(r[4]) for r in rows]
    highs = [to_float(r[2]) for r in rows]
    lows = [to_float(r[3]) for r in rows]
    vols = [to_float(r[5]) for r in rows]
    last, first = closes[-1], closes[0]
    if not first:
        return None

    ret = (last - first) / first * 100
    ret_3 = (closes[-1] - closes[-4]) / closes[-4] * 100 if len(closes) > 4 and closes[-4] else 0.0

    rsi = wilder_rsi(closes, 14)
    vol_last = vols[-1] if vols else 0.0
    vol_med = statistics.median(vols[:-1]) if len(vols) > 1 else 1.0

    trs = []
    for i, r in enumerate(rows):
        if i == 0:
            tr = to_float(r[2]) - to_float(r[3])
        else:
            pc = to_float(rows[i - 1][4])
            tr = max(to_float(r[2]) - to_float(r[3]), abs(to_float(r[2]) - pc), abs(to_float(r[3]) - pc))
        trs.append(tr)
    atr = statistics.mean(trs[-14:]) if len(trs) >= 14 else statistics.mean(trs)
    atr_pct = atr / last * 100 if last else 0.0

    if len(closes) >= 20:
        sma = statistics.mean(closes[-20:])
        sd = statistics.stdev(closes[-20:])
        bb_pct = (last - sma) / (2 * sd) * 100 if sd else 0.0
    else:
        bb_pct = 0.0

    if len(closes) >= 35:
        ema12 = _ema(closes, 12)
        ema26 = _ema(closes, 26)
        macd_series = [a - b for a, b in zip(ema12, ema26)]
        signal_series = _ema(macd_series[-26:] if len(macd_series) >= 26 else macd_series, 9)
        macd_line, signal_line = macd_series[-1], signal_series[-1]
        histogram = macd_line - signal_line
    else:
        macd_line = signal_line = histogram = 0.0

    if len(closes) >= 14:
        hh, ll = max(highs[-14:]), min(lows[-14:])
        stoch_rsi = (last - ll) / (hh - ll) * 100 if hh != ll else 50.0
    else:
        stoch_rsi = 50.0

    typical = [(to_float(r[2]) + to_float(r[3]) + to_float(r[4])) / 3 for r in rows]
    vol_sum = sum(vols)
    vwap = sum(t * v for t, v in zip(typical, vols)) / vol_sum if vol_sum else last

    support_tests = resistance_tests = 0
    if len(closes) > 5:
        for i in range(1, 5):
            if abs(lows[-i] - lows[-i - 1]) / (lows[-i] + 1e-6) < 0.01:
                support_tests += 1
            if abs(highs[-i] - highs[-i - 1]) / (highs[-i] + 1e-6) < 0.01:
                resistance_tests += 1

    return {
        "label": label, "ret": ret, "ret_3": ret_3, "rsi": rsi, "atr": atr, "atr_pct": atr_pct,
        "bb_pct": bb_pct, "last": last, "high": max(highs), "low": min(lows),
        "vol_total": sum(vols), "n": len(closes),
        "macd_line": macd_line, "signal_line": signal_line, "histogram": histogram,
        "stoch_rsi": stoch_rsi, "vwap": vwap,
        "support_tests": support_tests, "resistance_tests": resistance_tests,
    }

File: android/app/src/main/java/com/coinglass/intel/domain/Indicators.kt
package com.coinglass.intel.domain

import com.coinglass.intel.domain.model.Candle
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.sqrt

object Indicators {
    private fun ema(data: List<Double>, span: Int): List<Double> {
        if (data.isEmpty()) return emptyList()
        val alpha = 2.0 / (span + 1)
        val out = mutableListOf(data.first())
        for (i in 1 until data.size) {
            out += data[i] * alpha + out.last() * (1 - alpha)
        }
        return out
    }

    private fun stdev(xs: List<Double>): Double {
        if (xs.size < 2) return 0.0
        val m = xs.average()
        val v = xs.sumOf { (it - m) * (it - m) } / (xs.size - 1)
        return sqrt(v)
    }

    fun candleMetrics(rows: List<Candle>, label: String = ""): Map<String, Double>? {
        if (rows.size < 5) return null
        val sorted = rows.sortedBy { it.openTime }
        val closes = sorted.map { it.close }
        val highs = sorted.map { it.high }
        val lows = sorted.map { it.low }
        val vols = sorted.map { it.volume }
        val last = closes.last()
        val first = closes.first()
        if (first == 0.0) return null
        val ret = (last - first) / first * 100.0
        val ret3 = if (closes.size > 4 && closes[closes.size - 4] != 0.0) {
            (closes.last() - closes[closes.size - 4]) / closes[closes.size - 4] * 100.0
        } else 0.0
        val rsi = Scalper.wilderRsi(closes, 14)
        val volLast = vols.lastOrNull() ?: 0.0
        val volMed = if (vols.size > 1) {
            val body = vols.dropLast(1).sorted()
            body[body.size / 2]
        } else 1.0

        val trs = sorted.mapIndexed { i, r ->
            if (i == 0) r.high - r.low
            else {
                val pc = sorted[i - 1].close
                max(r.high - r.low, max(abs(r.high - pc), abs(r.low - pc)))
            }
        }
        val atr = if (trs.size >= 14) trs.takeLast(14).average() else trs.average()
        val atrPct = if (last == 0.0) 0.0 else atr / last * 100.0

        val bbPct = if (closes.size >= 20) {
            val win = closes.takeLast(20)
            val sma = win.average()
            val sd = stdev(win)
            if (sd == 0.0) 0.0 else (last - sma) / (2 * sd) * 100.0
        } else 0.0

        var macdLine = 0.0
        var signalLine = 0.0
        var histogram = 0.0
        if (closes.size >= 35) {
            val ema12 = ema(closes, 12)
            val ema26 = ema(closes, 26)
            val macdSeries = ema12.zip(ema26) { a, b -> a - b }
            val signalSeries = ema(if (macdSeries.size >= 26) macdSeries.takeLast(26) else macdSeries, 9)
            macdLine = macdSeries.last()
            signalLine = signalSeries.last()
            histogram = macdLine - signalLine
        }

        val stochRsi = if (closes.size >= 14) {
            val hh = highs.takeLast(14).max()
            val ll = lows.takeLast(14).min()
            if (hh == ll) 50.0 else (last - ll) / (hh - ll) * 100.0
        } else 50.0

        val typical = sorted.map { (it.high + it.low + it.close) / 3.0 }
        val volSum = vols.sum()
        val vwap = if (volSum == 0.0) last else typical.zip(vols) { t, v -> t * v }.sum() / volSum

        var supportTests = 0.0
        var resistanceTests = 0.0
        if (closes.size > 5) {
            for (i in 1..4) {
                val li = lows[lows.size - i]
                val li1 = lows[lows.size - i - 1]
                val hi = highs[highs.size - i]
                val hi1 = highs[highs.size - i - 1]
                if (abs(li - li1) / (li + 1e-6) < 0.01) supportTests += 1
                if (abs(hi - hi1) / (hi + 1e-6) < 0.01) resistanceTests += 1
            }
        }

        return mapOf(
            "ret" to ret,
            "ret_3" to ret3,
            "rsi" to rsi,
            "atr" to atr,
            "atr_pct" to atrPct,
            "bb_pct" to bbPct,
            "last" to last,
            "high" to (highs.maxOrNull() ?: 0.0),
            "low" to (lows.minOrNull() ?: 0.0),
            "vol_total" to vols.sum(),
            "vol_last" to volLast,
            "vol_med" to volMed,
            "n" to closes.size.toDouble(),
            "macd_line" to macdLine,
            "signal_line" to signalLine,
            "histogram" to histogram,
            "stoch_rsi" to stochRsi,
            "vwap" to vwap,
            "support_tests" to supportTests,
            "resistance_tests" to resistanceTests,
        )
    }
}
